### PR TITLE
Support for url encoded parameters and httpBody

### PR DESCRIPTION
**Pull request by PChmiel**

### CODE REVIEW

File: C0T0/Classes/ApiDataSource/URLRequestBuilder.swift [hunk 1]
```diff
@@ -18,12 +18,21 @@ class StandardURLRequestBuilder: URLRequestBuilder {
     }
 
     func build(from request: ApiRequest) -> URLRequest? {
-        guard let  url = URL(string: host + request.endpoint) else { return nil }
+        var urlParametersString = ""
+        if let urlParameters = request.urlParameters {
```
Comment (Vyeczorny) [inline C0T0/Classes/ApiDataSource/URLRequestBuilder.swift]: A bit more concise code for that:
```
let urlParametersString = urlParameters?.map { "\(key)=\(value)" }.joined(separator: "&") ?? ""
```

File: C0T0/Classes/ApiDataSource/URLRequestBuilder.swift [hunk 2]
```diff
@@ -18,12 +18,21 @@ class StandardURLRequestBuilder: URLRequestBuilder {
...
+            for (key, value) in urlParameters {
+                urlParametersString += "&\(key)=\(value)"
+            }
+        }
+
+        guard let  url = URL(string: host + request.endpoint + urlParametersString) else { return nil }
```
Comment (Vyeczorny) [inline C0T0/Classes/ApiDataSource/URLRequestBuilder.swift]: I think there is a small error. Current implementation adds `&` before every parameter, so before first parameter too. imo the sign which separates endpoint and parameters is `?`, not `&`.

File: C0T0/Classes/Models/ApiRequest.swift [hunk 3]
```diff
@@ -6,6 +6,7 @@ import Foundation
 
 public typealias HTTPParameters = [String: Any]
 public typealias HTTPHeaders = [String: String]
+public typealias HTTPBody = [String: Any]
```
Comment (Vyeczorny) [inline C0T0/Classes/Models/ApiRequest.swift]: I think `HTTPBody` should be just `Data`

File: C0T0/Classes/Models/ApiRequest.swift [hunk 4]
```diff
@@ -23,33 +24,45 @@ public struct ApiRequest {
...
+        self.httpBody = httpBody
     }
 }
 
 extension ApiRequest: Equatable {
 
     public static func == (lhs: ApiRequest, rhs: ApiRequest) -> Bool {
         let equalParameters: Bool
-        switch (lhs.parameters, rhs.parameters) {
+        switch (lhs.urlParameters, rhs.urlParameters) {
```
Comment (Vyeczorny) [inline C0T0/Classes/Models/ApiRequest.swift]: Looks like nice code for separate comparing function ;)

**Review by Vyeczorny — CHANGES_REQUESTED** [review]

File: C0T0/Classes/ApiDataSource/URLRequestBuilder.swift [hunk 5]
```diff
@@ -18,12 +18,21 @@ class StandardURLRequestBuilder: URLRequestBuilder {
     }
 
     func build(from request: ApiRequest) -> URLRequest? {
-        guard let  url = URL(string: host + request.endpoint) else { return nil }
+        var urlParametersString = ""
+        if let urlParameters = request.urlParameters {
+            for (key, value) in urlParameters {
```
Comment (pstasiak) [inline C0T0/Classes/ApiDataSource/URLRequestBuilder.swift]: Why not constructing url using `URLComponents.queryItems` + 'URLQueryItem'?

**Review by pstasiak — APPROVED** [review]

File: C0T0/Classes/ApiDataSource/URLRequestBuilder.swift [hunk 6]
```diff
@@ -10,20 +10,41 @@ protocol URLRequestBuilder {
 }
 
 class StandardURLRequestBuilder: URLRequestBuilder {
-
+    private enum Constants {
```
Comment (Vyeczorny) [inline C0T0/Classes/ApiDataSource/URLRequestBuilder.swift]: I must admit I don't like this enum. Will we use `slashes` constant without scheme? If not, U don't see any advantages of keeping separate constants and I would leave just two constants: `http://` and `https://`.

File: C0T0/Classes/ApiDataSource/URLRequestBuilder.swift [hunk 7]
```diff
@@ -10,20 +10,41 @@ protocol URLRequestBuilder {
...
+    }
+    
     private let host: String
-
+    private let scheme: String
     init(host: String) {
+        var host = host
+        var scheme = ""
+        if host.contains(Constants.httpProtocol) {
+            host = host.replacingOccurrences(of:Constants.httpProtocol , with: "")
```
Comment (Vyeczorny) [inline C0T0/Classes/ApiDataSource/URLRequestBuilder.swift]: You can just make an assignment like: `self.host = host.replacingOccurrences(of:Constants.httpProtocol , with: "")`. I will allow you to remove two local variables and four lines of code :)

Comment (Vyeczorny) [reply C0T0/Classes/ApiDataSource/URLRequestBuilder.swift]: Btw, you missed space after `:` :)

File: C0T0/Classes/ApiDataSource/URLRequestBuilder.swift [hunk 8]
```diff
@@ -10,20 +10,41 @@ protocol URLRequestBuilder {
...
+            scheme = Constants.https
+        }
         self.host = host
+        self.scheme = scheme
     }
 
     func build(from request: ApiRequest) -> URLRequest? {
-        guard let  url = URL(string: host + request.endpoint) else { return nil }
-        var urlRequest = URLRequest(url: url)
+        var components = URLComponents()
```
Comment (Vyeczorny) [inline C0T0/Classes/ApiDataSource/URLRequestBuilder.swift]: image
As you can see on the screenshot, `URLComponents` can split address into scheme and host, so you don't need to do it manually, just pass whole `host` to it :)

**Review by Vyeczorny — CHANGES_REQUESTED** [review]

**Review by Vyeczorny — APPROVED** [review]

**Pull request merged**